default/python/AI/SaveGameManager/_decoder.py

```diff
@@ -0,0 +1,181 @@
+"""This module defines the decoding for the FreeOrion AI savegames.
+
+The decoder is subclassed from the standard library json decoder and
+uses its string parsing. However, the resulting objects will be interpreted
+differently according to the encoding used in FreeOrion AI savegames.
+
+The decoder will only load trusted classes as defined in _definitions.py,
+if an unknown/untrusted object is encountered, it will raise a InvalidSaveGameException.
+
+When classes are loaded, their __setstate__ method will be invoked if available or
+the __dict__ content will be set directly. It is the responsiblity of the trusted classes
+to provide a __setstate__ method to verify and possibly sanitize the content of the passed state.
+
+Example usage:
+    Importing this module directly is not advised.
+    Use the exported functions when importing the package instead.
+"""
+import json
+
+import EnumsAI
+from freeorion_tools import profile
+
+from _definitions import *
+
+
+@profile
+def load_savegame_string(string):
+    import zlib
+    try:
+        string = zlib.decompress(string)
+    except zlib.error:
+        pass  # probably an uncompressed string
+    return decode(string)
+
+
+def decode(obj):
+    return _FreeOrionAISaveGameDecoder().decode(obj)
+
+
+class _FreeOrionAISaveGameDecoder(json.JSONDecoder):
+
+    def __init__(self, **kwargs):
+        super(_FreeOrionAISaveGameDecoder, self).__init__(strict=True,  # do not allow control characters
+                                                          **kwargs)
+
+    def decode(self, s, _w=None):
+        # use the default JSONDecoder to parse the string into a dict
+        # then interpret the dict content according to our encoding
+        retval = super(_FreeOrionAISaveGameDecoder, self).decode(s)
+        return self.__interpret(retval)
+
+    def __interpret_dict(self, obj):
+        # With our encoding, dicts with more than one entry must be a standard dict
+        # because our custom encoded classes have only a single key which is its name
+        # and its __dict__ content as value. In that case, interpret the dict content.
+        if len(obj) != 1:
```

Comment:
In java jackson library next style is used:

Base item can be `int`, `float`, `string` and `array`.   `array` has always two elements, first is class name string and second is some value (including dict), representing this class.
Example:
`1` -> `1`
`['list', [1, 2, 3]]` - > `[1, 2, 3]`
`["dict", {"hello": ["dict", {"1": 2}]}]` - > `{'hello': {'1': 2}}`

Reply:
How is ```["Type", value]``` preferable over ```{"Type": value}```?

Reply:
> With our encoding, dicts with more than one entry must be a standard dict

`{'boo': 'foo'}` a dict or not a dict?   
`['dict': `{'boo': 'foo'}`]`  its  a dict!

```python
import this
'In the face of ambiguity, refuse the temptation to guess.'
```

Reply:
Seems to me that the current plan/requirement to implement setstate/getstate for custom classes pretty much does away with this concern about nonstandard dicts, at worst we just have to make an FO-flavored version of the nonstandard dict and implement setstate and getstate for it.

Reply:
Can you rephrase.  I don't sure that I understand what you mean.

I see issue, that I can make a simple dict, that will be serialized to file and will break deserialization.

`{CLASS_PREFIX: 'try to save me'}`

Reply:
> I see issue, that I can make a simple dict, that will be serialized to file and will break deserialization.
Using the PLACEHOLDER char in strings within sets/tuples would also break deserialization.

I doubt that either prefix or PLACEHOLDER char ```$``` would be used in practice.


> In java jackson library next style is used:

I tried to look this up but all output for the jackson library I found was standard json which does not have the style you outlined here. Do you have any links for me?

The suggested style seems easy enough to break on purpose as well unless I am overlooking something...
```
decode(encode(['list', [1, 2, 3]]))
-->decode( ['list', ['list', ['list', [1, 2, 3]]])
--> [[[1,2,3]]]
```
You would probably have to encode strings as well to avoid ambiguity:
```
decode(encode(['list', [1, 2, 3]]))
-->decode( ['list', [['string', 'list'], ['list', [1, 2, 3]]])
--> ['list', [1,2,3]]
```

Reply:
> I doubt that either prefix or PLACEHOLDER char $ would be used in practice.

`this will never happens` usually happens more often when never.   Lets implement: `this will never happen`

>  Do you have any links for me?

I know that we use custom json mapper on my job, but did not look deep inside yet, I will try to not forget to check how it is setuped and try to find any url about it. 

Meantime will try to explain this by myself:

On top level we always have class. (list of two elements string, json object)

So we just find suitable class and pass data to it. Class will know how to work with it.

In my example:
```python
class AIFleetOrder(Model):
    fleet = FieldClass(Fleet, may_be_none=False)
    target = FieldClass(TARGET_TYPE)
    executed = FieldBoolean()
    order_issued = FieldBoolean()
```

```python
['AIFleetOrder', 
    {
        'fleet': ['Fleet', {....}],
        'target': [{'target_class', {...}], 
        'executed': True,  
        'order_issued': True 
    }
]
```

Little redundancy, because we have `Fleet` in declaration and in serializable object, on of them should be removed.